### DISCUSSION TITLE
Calling save_changes multiple times on a reference produces an error.  Work around?  Fix?

### DISCUSSION THREAD

**Discussion by randall-coding** [diesel-rs/diesel#2816] [Q&A]
I'm adding retrys to my update calls to retry if there is a locked database error, but I'm noticing if `save_changes` is called multiple times on the same reference I get this error.  I don't fully understand what this error message is telling me.

>  the trait bound `std::option::Option<models::Upload>: Queryable<(diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Double, diesel::sql_types::Double, diesel::sql_types::Bool, diesel::sql_types::Bool, diesel::sql_types::Nullable<diesel::sql_types::Date>, diesel::sql_types::Text), Sqlite>` is not satisfied
> 
> the following implementations were found: <std::option::Option<T> as Queryable<diesel::sql_types::Nullable<ST>, DB>>
> required because of the requirements on the impl of `LoadQuery<diesel::SqliteConnection, std::option::Option<models::Upload>>` for `diesel::query_builder::SelectStatement<schema::uploads::table, diesel::query_builder::select_clause::DefaultSelectClause, diesel::query_builder::distinct_clause::NoDistinctClause, diesel::query_builder::where_clause::WhereClause<diesel::expression::operators::Eq<schema::uploads::columns::id, diesel::expression::bound::Bound<diesel::sql_types::Text, &std::string::String>>>>`
> required because of the requirements on the impl of `UpdateAndFetchResults<&models::Upload, std::option::Option<models::Upload>>` for `diesel::SqliteConnection

This code works in my previous code with a single save_changes

```
 pub fn update(obj: &Upload) -> Option<$Upload> {
            let conn = establish_connection();
            let res = obj.save_changes(&conn) ;
            ...
       }
```

If I try the code below it yields the error I showed above.  This is a simplified example calling save_changed twice.  
```
 pub fn update(obj: &Upload, retrys: Option<i32>) -> Option<$Upload> {
            let conn = establish_connection();
            let res = obj.save_changes(&conn) ;  //error message now shows up
            let res2 = obj.save_changes(&conn) ;  // do this if res is an error
            ...
       }
```

Does anyone have insight into why this is happening and how I can make this work?

**Comment by weiznich** [comment]
It's currently impossible to answer this without having more information. Please provide at least the following information:
* The complete definition of `update` with concrete types
* All definitions of involved user defined types (So at least `model::Upload` in this case), including all relevant derives
* The corresponding `table!` definitions from your schema.rs
* Information about the used compiler and diesel version
* The complete error message formatted as code block

**Reply by randall-coding** [reply]
Thanks, I'll get that information together.

**Comment by randall-coding** [comment]
@weiznich Here is my model::Upload  

```
#[derive(Serialize, Queryable, Identifiable, AsChangeset, Debug, Clone)]
#[changeset_options(treat_none_as_null="true")] 
pub struct Upload {
    pub id: String, 
    pub video_id: String,
    pub user_id: String,
    pub platform: String,
    pub downloaded_size: f64,
    pub total_size: f64,
    pub completed: bool,
    pub paused: bool,
    pub created_at: Option<String>,
    pub status: String,
}

impl Upload {
    add_formatted_date!();  //some boiler plate code put into macros
    add_formatted_size!(downloaded_size);
    add_formatted_size!(total_size);
}

#[derive(Deserialize, Insertable)]
#[table_name = "uploads"]
pub struct NewUpload<'a> {
    pub id: &'a str,
    pub video_id: &'a str,
    pub user_id: &'a str,
    pub platform: &'a str,    
    pub created_at: Option<&'a str>,
}
```

Here is the schema for Upload 

```
table! {
    uploads (id) {
        id -> Text,
        video_id -> Text,
        user_id -> Text,
        platform -> Text,
        downloaded_size -> Double,
        total_size -> Double,
        completed -> Bool,
        paused -> Bool,
        created_at -> Nullable<Date>,
        status -> Text,
    }
}
```

Compiler `rustc 1.53.0-nightly (d32238532 2021-04-05)` 
Diesel crate `diesel = { version = "1.4.5", features = ["sqlite"] }`
Diesel CLI `diesel 1.4.1` 

Here is the error message showing up on save_changes

```
the trait bound `std::option::Option<models::Upload>: Queryable<(diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Double, diesel::sql_types::Double, diesel::sql_types::Bool, diesel::sql_types::Bool, diesel::sql_types::Nullable<diesel::sql_types::Date>, diesel::sql_types::Text), Sqlite>` is not satisfied
the following implementations were found:
  <std::option::Option<T> as Queryable<diesel::sql_types::Nullable<ST>, DB>>
required because of the requirements on the impl of `LoadQuery<diesel::SqliteConnection, std::option::Option<models::Upload>>` for `diesel::query_builder::SelectStatement<schema::uploads::table, diesel::query_builder::select_clause::DefaultSelectClause, diesel::query_builder::distinct_clause::NoDistinctClause, diesel::query_builder::where_clause::WhereClause<diesel::expression::operators::Eq<schema::uploads::columns::id, diesel::expression::bound::Bound<diesel::sql_types::Text, &std::string::String>>>>`
required because of the requirements on the impl of `UpdateAndFetchResults<&models::Upload, std::option::Option<models::Upload>>` for `diesel::SqliteConnection`rustcE0277
```

Here is my update function test in full.  I forgot to add the type annotation last time (when I call save_changes twice the compiler insists I add the type annotation on the result).  

```
pub fn update(obj: &Upload) -> Option<Upload> {
    let conn = establish_connection();
    
    let res: Result<Option<Upload>, diesel::result::Error> = obj.save_changes(&conn);
    let res = obj.save_changes(&conn) ;
    
    if res.is_err(){
        let err = res.err().unwrap();
        errormsg::message!(err);
        return None
    }
    return Some(res.unwrap())
}
```

**Comment by randall-coding** [comment]
My original attempt at the retrys didn't even have the type annotation but gave the same error

```
pub fn update(obj: &Upload, retrys_opt: Option<i32>) -> Option<Upload> {
    let retrys = match retrys_opt {
        Some(retrys) => {retrys}
        None => {1}
    };
    let conn = establish_connection();
    let res = obj.save_changes(&conn);
 
    let mut retry_counter = 0; 
    let output: Option<Upload> = None; 
    while retry_counter < retrys {
        retry_counter += 1;
        println!("retry_counter {}", retry_counter);
        if retry_counter > 1 {
            sleep(Duration::from_millis(100));
        }

        let conn = establish_connection();
        let res = obj.save_changes(&conn);  // error here 
        ...
```

**Comment by randall-coding** [comment]
@weiznich  

Looks like I resolved it!  Adding `Result<Upload, diesel::result::Error>` as an annotation to the result stops the error (while using `Result<Option<Upload>, diesel::result::Error>` didn't.  
 
I found it by playing around with that simplified test I showed above.  Still it's puzzling why calling it twice requires explicit type annotation but once doesn't.

**Reply by weiznich** [reply]
The type annotation is required there because you just don't use the first `res` anywhere. It's a completely different variable to the second `res`, that just happens to share the same name.

**Reply by randall-coding** [reply]
Oh so if the `res` isn't used the compiler isn't able to infer the result type(?).  I think I understand now.  

Now that you mention it that line didn't need to be there anyway, so it's been removed, thanks!
